File: pipeline/src/workers/source_scout.py
"""
Source Scout pipeline worker.

Crawls source_profiles with auto_crawl_enabled, compares content hashes
for each annotated region, calls Claude for analysis on changed regions,
and emits events for downstream workflows.

Designed to run as a pipeline_jobs kind='scout_source' handler.
Can scout a single source (by metadata.source_id) or all due sources.
"""
from __future__ import annotations

import hashlib
import json
import logging
from datetime import datetime, timezone
from typing import Any, Optional

import asyncpg
import httpx

from events import emit_event

log = logging.getLogger("pipeline.workers.source_scout")

# Claude model for analysis
CLAUDE_MODEL = "claude-sonnet-4-20250514"
MAX_CONTENT_LENGTH = 50_000
MAX_PROMPT_CONTENT = 12_000
MAX_PREVIOUS_CONTENT = 8_000

# Lazy-loaded anthropic client — tests override via ANTHROPIC_CLIENT attribute
ANTHROPIC_CLIENT = None


def _content_hash(text: str) -> str:
    """SHA-256 hash of text content for change detection."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _html_to_text(html: str) -> str:
    """Strip HTML tags to plain text for hashing and comparison."""
    import re

    text = re.sub(r"<script[^>]*>[\s\S]*?</script>", "", html, flags=re.IGNORECASE)
    text = re.sub(r"<style[^>]*>[\s\S]*?</style>", "", text, flags=re.IGNORECASE)
    text = re.sub(r"<[^>]+>", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


async def _fetch_page(url: str) -> tuple[str, Optional[str]]:
    """Fetch page HTML via httpx. Returns (html, warning_or_none)."""
    async with httpx.AsyncClient(
        timeout=30.0,
        follow_redirects=True,
        headers={
            "User-Agent": "GovWin-SourceScout/1.0 (Federal opportunity monitoring)",
            "Accept": "text/html,application/xhtml+xml",
        },
    ) as client:
        resp = await client.get(url)
        resp.raise_for_status()
        html = resp.text

    # Heuristic: JS-only page detection
    import re

    body_match = re.search(r"<body[^>]*>([\s\S]*?)</body>", html, re.IGNORECASE)
    body_text = re.sub(r"<[^>]+>", "", body_match.group(1)).strip() if body_match else ""
    warning = None
    if len(body_text) < 100 and "__NEXT_DATA__" in html:
        warning = (
            "This site appears to require JavaScript rendering. "
            "Content may be incomplete."
        )
    return html, warning


async def _analyze_with_claude(
    source_name: str,
    region_name: str,
    content_context: str,
    previous_text: Optional[str],
    current_text: str,
) -> Optional[dict[str, Any]]:
    """Ask Claude to analyze a region for changes. Returns parsed analysis or None."""
    global ANTHROPIC_CLIENT

    try:
        import anthropic
    except ImportError:
        log.warning("anthropic SDK not available — skipping Claude analysis")
        return None

    if ANTHROPIC_CLIENT is None:
        ANTHROPIC_CLIENT = anthropic.AsyncAnthropic()

    previous_section = (
        f"<previous_content>\n{previous_text[:MAX_PREVIOUS_CONTENT]}\n</previous_content>"
        if previous_text
        else "<previous_content>No previous snapshot available — this is the first scan.</previous_content>"
    )

    prompt = f"""You are analyzing a government opportunity website for changes.

Source: {source_name}
Region: "{region_name}"
Admin guidance: "{content_context or 'No specific guidance provided.'}"

{previous_section}

<current_content>
{current_text[:MAX_PROMPT_CONTENT]}
</current_content>

Analyze the current content for this region. Respond in JSON only:

{{
  "changed": true/false,
  "summary": "Brief description of what changed or 'No changes detected'",
  "severity": "info|low|medium|high|critical",
  "extracted_opportunities": [
    {{
      "title": "...",
      "agency": "...",
      "deadline": "...",
      "url": "...",
      "description": "..."
    }}
  ]
}}

Severity guide:
- info: no meaningful change, cosmetic only
- low: minor text updates, no new opportunities
- medium: updated deadlines or modified requirements
- high: new opportunities or solicitations detected
- critical: imminent deadlines or major program changes

Only include extracted_opportunities if you identify specific, actionable opportunities."""

    try:
        response = await ANTHROPIC_CLIENT.messages.create(
            model=CLAUDE_MODEL,
            max_tokens=2048,
            messages=[{"role": "user", "content": prompt}],
        )

        text = response.content[0].text if response.content else ""
        tokens_used = (response.usage.input_tokens or 0) + (response.usage.output_tokens or 0)

        import re

        json_match = re.search(r"\{[\s\S]*\}", text)
        if not json_match:
            return {
                "changed": False,
                "summary": "Claude response did not contain valid JSON",
                "severity": "info",
                "extracted_opportunities": [],
                "model": response.model,
                "tokens_used": tokens_used,
            }

        parsed = json.loads(json_match.group(0))
        return {
            "changed": parsed.get("changed", False),
            "summary": parsed.get("summary", "No summary provided"),
            "severity": parsed.get("severity", "info"),
            "extracted_opportunities": parsed.get("extracted_opportunities", []),
            "model": response.model,
            "tokens_used": tokens_used,
        }
    except Exception as e:
        log.error("Claude analysis failed: %s", e)
        return None


async def scout_source(
    conn: asyncpg.Connection,
    source_id: str,
) -> dict[str, Any]:
    """Scout a single source profile for changes.

    Returns a result dict with counts of snapshots, diffs, and changes.
    """
    # Load source profile
    profile = await conn.fetchrow(
        """
        SELECT id, name, base_url, bookmark_url, is_active
        FROM source_profiles
        WHERE id = $1
        """,
        source_id,
    )
    if not profile:
        log.warning("source profile not found: %s", source_id)
        return {"error": f"source profile not found: {source_id}"}

    # Load active regions
    regions = await conn.fetch(
        """
        SELECT id, name, selector_hint, content_context, region_type, sample_text
        FROM source_regions
        WHERE profile_id = $1 AND is_active = true
        ORDER BY created_at
        """,
        source_id,
    )

    warnings: list[str] = []

    # Fetch the page
    fetch_url = profile["bookmark_url"] or profile["base_url"]
    try:
        page_html, warning = await _fetch_page(fetch_url)
        if warning:
            warnings.append(warning)
    except Exception as e:
        log.error("failed to fetch %s: %s", fetch_url, e)
        return {"error": f"failed to fetch page: {e}", "warnings": warnings}

    full_page_text = _html_to_text(page_html)

    # Process regions (or full page if no regions)
    snapshots_created = 0
    diffs_found = 0
    meaningful_changes = 0
    region_results: list[dict[str, Any]] = []

    if not regions:
        regions = [
            {
                "id": None,
                "name": "Full Page",
                "selector_hint": None,
                "content_context": "Full page content",
                "region_type": "content",
                "sample_text": None,
            }
        ]
        warnings.append("No active regions defined — scanning full page")

    for region in regions:
        region_id = region["id"]
        region_text = full_page_text
        hash_val = _content_hash(region_text)

        # Get last snapshot
        if region_id:
            last_snapshot = await conn.fetchrow(
                """
                SELECT id, content_hash, content_text
                FROM source_snapshots
                WHERE profile_id = $1 AND region_id = $2
                ORDER BY captured_at DESC
                LIMIT 1
                """,
                source_id,
                region_id,
            )
        else:
            last_snapshot = await conn.fetchrow(
                """
                SELECT id, content_hash, content_text
                FROM source_snapshots
                WHERE profile_id = $1 AND region_id IS NULL
                ORDER BY captured_at DESC
                LIMIT 1
                """,
                source_id,
            )

        hash_changed = not last_snapshot or last_snapshot["content_hash"] != hash_val

        # Create new snapshot
        new_snapshot = await conn.fetchrow(
            """
            INSERT INTO source_snapshots
              (profile_id, region_id, content_hash, content_text)
            VALUES ($1, $2, $3, $4)
            RETURNING id
            """,
            source_id,
            region_id,
            hash_val,
            region_text[:MAX_CONTENT_LENGTH],
        )
        snapshots_created += 1
        new_snapshot_id = new_snapshot["id"]

        if not hash_changed:
            region_results.append({
                "regionId": str(region_id) if region_id else "full-page",
                "regionName": region["name"],
                "changed": False,
                "summary": "No changes detected (content hash match)",
                "severity": "info",
            })
            continue

        # Hash changed — analyze with Claude
        analysis = await _analyze_with_claude(
            profile["name"],
            region["name"],
            region["content_context"] or "",
            (last_snapshot["content_text"] if last_snapshot else None) or region["sample_text"],
            region_text,
        )

        is_meaningful = analysis["changed"] if analysis else hash_changed
        summary = analysis["summary"] if analysis else "Content hash changed but Claude analysis unavailable"
        severity = analysis["severity"] if analysis else "low"
        # source_diffs.severity is CHECK-constrained; an LLM-returned value outside the
        # set would abort the whole scout run at INSERT (not per-row wrapped). Clamp it.
        if severity not in ("info", "low", "medium", "high", "critical"):
            severity = "medium"
        extracted_opps = analysis.get("extracted_opportunities", []) if analysis else []

        # Create diff record
        await conn.execute(
            """
            INSERT INTO source_diffs
              (profile_id, region_id, prev_snapshot_id, next_snapshot_id,
               is_meaningful, summary, extracted_opportunities, severity,
               claude_model, claude_tokens_used)
            VALUES ($1, $2, $3, $4, $5, $6, $7::jsonb, $8, $9, $10)
            """,
            source_id,
            region_id,
            last_snapshot["id"] if last_snapshot else None,
            new_snapshot_id,
            is_meaningful,
            summary,
            json.dumps(extracted_opps),
            severity,
            analysis.get("model") if analysis else None,
            analysis.get("tokens_used") if analysis else None,
        )

        diffs_found += 1
        if is_meaningful:
            meaningful_changes += 1

        region_results.append({
            "regionId": str(region_id) if region_id else "full-page",
            "regionName": region["name"],
            "changed": is_meaningful,
            "summary": summary,
            "severity": severity,
            "extractedOpportunities": extracted_opps,
        })

    # Update last_crawl_at
    await conn.execute(
        "UPDATE source_profiles SET last_crawl_at = now(), updated_at = now() WHERE id = $1",
        source_id,
    )

    # Emit events
    await emit_event(
        conn,
        namespace="finder",
        type="source.scouted",
        payload={
            "sourceId": str(source_id),
            "sourceName": profile["name"],
            "snapshotsCreated": snapshots_created,
            "diffsFound": diffs_found,
            "meaningfulChanges": meaningful_changes,
            "warnings": warnings,
        },
    )

    if meaningful_changes > 0:
        await emit_event(
            conn,
            namespace="finder",
            type="source.change_detected",
            payload={
                "sourceId": str(source_id),
                "sourceName": profile["name"],
                "meaningfulChanges": meaningful_changes,
                "regionResults": [r for r in region_results if r.get("changed")],
            },
        )

    result = {
        "sourceId": str(source_id),
        "sourceName": profile["name"],
        "snapshotsCreated": snapshots_created,
        "diffsFound": diffs_found,
        "meaningfulChanges": meaningful_changes,
        "warnings": warnings,
    }

    log.info(
        "scout completed for %s: snapshots=%d diffs=%d meaningful=%d",
        profile["name"],
        snapshots_created,
        diffs_found,
        meaningful_changes,
    )

    return result


async def scout_all_due(conn: asyncpg.Connection) -> dict[str, Any]:
    """Scout all source profiles where auto_crawl is enabled and crawl is due.

    A crawl is due when last_crawl_at is NULL or older than the crawl_cron
    interval would dictate. For V1, we use a simple 24-hour check.
    """
    profiles = await conn.fetch(
        """
        SELECT id FROM source_profiles
        WHERE auto_crawl_enabled = true
          AND is_active = true
          AND (last_crawl_at IS NULL
               OR last_crawl_at < now() - INTERVAL '23 hours')
        ORDER BY last_crawl_at ASC NULLS FIRST
        """
    )

    results = []
    for profile in profiles:
        try:
            result = await scout_source(conn, profile["id"])
            results.append(result)
        except Exception as e:
            log.error("scout failed for source %s: %s", profile["id"], e)
            results.append({"sourceId": str(profile["id"]), "error": str(e)})

    return {
        "sourcesChecked": len(profiles),
        "results": results,
    }

File: frontend/lib/tools/source-scout.ts
/**
 * finder.scout_source — Source Scout tool.
 *
 * HITL-guided web monitoring: fetches a source page, compares content
 * against admin-annotated regions, and uses Claude to detect meaningful
 * changes (new opportunities, updated topics, deadline changes).
 *
 * Callable by:
 *   - Admin UI "Scout Now" button (manual trigger)
 *   - Pipeline worker (automated cron via pipeline_jobs kind='scout_source')
 *
 * V1 approach: plain HTTP fetch (no Playwright). Government sites
 * render server-side. If JS rendering is needed, the tool returns a
 * warning and the admin falls back to manual paste.
 */

import { z } from 'zod';
import { createHash, randomUUID } from 'crypto';
import { sql } from '@/lib/db';
import { emitEventSingle } from '@/lib/events';
import { NotFoundError } from '@/lib/errors';
import { ToolExternalError } from './errors';
import { defineTool } from './base';

// ─── Schema ──────────────────────────────────────────────────────────

const InputSchema = z.object({
  sourceId: z.string().uuid(),
});

type Input = z.infer<typeof InputSchema>;

interface RegionDiff {
  regionId: string;
  regionName: string;
  changed: boolean;
  summary: string | null;
  severity: string;
  extractedOpportunities: unknown[];
}

interface Output {
  sourceId: string;
  sourceName: string;
  snapshotsCreated: number;
  diffsFound: number;
  meaningfulChanges: number;
  regionResults: RegionDiff[];
  warnings: string[];
}

// ─── Helpers ─────────────────────────────────────────────────────────

function contentHash(text: string): string {
  return createHash('sha256').update(text).digest('hex');
}

/**
 * Fetch page HTML via plain HTTP. Returns null + a warning string if
 * the fetch fails or the response looks like it requires JS rendering.
 */
async function fetchPageHtml(
  url: string,
): Promise<{ html: string; warning: string | null }> {
  const res = await fetch(url, {
    headers: {
      'User-Agent':
        'GovWin-SourceScout/1.0 (Federal opportunity monitoring; +https://govwin.io)',
      Accept: 'text/html,application/xhtml+xml',
    },
    signal: AbortSignal.timeout(30_000),
  });

  if (!res.ok) {
    throw new ToolExternalError(
      `failed to fetch ${url}: HTTP ${res.status} ${res.statusText}`,
      { url, status: res.status },
    );
  }

  const html = await res.text();

  // Heuristic: if the body is tiny and contains only a JS bootstrap,
  // the site likely requires client-side rendering.
  const bodyMatch = html.match(/<body[^>]*>([\s\S]*?)<\/body>/i);
  const bodyText = bodyMatch?.[1]?.replace(/<[^>]+>/g, '').trim() ?? '';
  if (bodyText.length < 100 && html.includes('__NEXT_DATA__')) {
    return {
      html,
      warning:
        'This site appears to require JavaScript rendering. Content may be incomplete. Consider using manual paste for this source.',
    };
  }

  return { html, warning: null };
}

/**
 * Strip HTML tags to produce plain text for hashing and comparison.
 */
function htmlToText(html: string): string {
  return html
    .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '')
    .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Ask Claude to analyze a page region for changes. Returns structured
 * analysis or null if Claude is unavailable.
 */
async function analyzeRegionWithClaude(
  sourceName: string,
  regionName: string,
  contentContext: string,
  previousText: string | null,
  currentText: string,
  currentHtml: string,
): Promise<{
  changed: boolean;
  summary: string;
  severity: string;
  extractedOpportunities: unknown[];
  model: string;
  tokensUsed: number;
} | null> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    return null;
  }

  const previousSection = previousText
    ? `<previous_content>\n${previousText.slice(0, 8000)}\n</previous_content>`
    : '<previous_content>No previous snapshot available — this is the first scan.</previous_content>';

  const prompt = `You are analyzing a government opportunity website for changes.

Source: ${sourceName}
Region: "${regionName}"
Admin guidance: "${contentContext ?? 'No specific guidance provided.'}"

${previousSection}

<current_content>
${currentText.slice(0, 12000)}
</current_content>

Analyze the current content for this region. Respond in JSON only:

{
  "changed": true/false,
  "summary": "Brief description of what changed or 'No changes detected'",
  "severity": "info|low|medium|high|critical",
  "extracted_opportunities": [
    {
      "title": "...",
      "agency": "...",
      "deadline": "...",
      "url": "...",
      "description": "..."
    }
  ]
}

Severity guide:
- info: no meaningful change, cosmetic only
- low: minor text updates, no new opportunities
- medium: updated deadlines or modified requirements
- high: new opportunities or solicitations detected
- critical: imminent deadlines or major program changes

Only include extracted_opportunities if you identify specific, actionable opportunities.`;

  try {
    const res = await fetch('https://api.anthropic.com/v1/messages', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': apiKey,
        'anthropic-version': '2023-06-01',
      },
      body: JSON.stringify({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 2048,
        messages: [{ role: 'user', content: prompt }],
      }),
      signal: AbortSignal.timeout(60_000),
    });

    if (!res.ok) {
      const errText = await res.text().catch(() => 'unknown');
      throw new Error(`Anthropic API error: ${res.status} — ${errText.slice(0, 200)}`);
    }

    const data = (await res.json()) as {
      content: { type: string; text: string }[];
      model: string;
      usage: { input_tokens: number; output_tokens: number };
    };
    const text = data.content?.[0]?.text ?? '';
    const tokensUsed =
      (data.usage?.input_tokens ?? 0) + (data.usage?.output_tokens ?? 0);

    // Parse the JSON from Claude's response
    const jsonMatch = text.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      return {
        changed: false,
        summary: 'Claude response did not contain valid JSON',
        severity: 'info',
        extractedOpportunities: [],
        model: data.model ?? 'unknown',
        tokensUsed,
      };
    }

    const parsed = JSON.parse(jsonMatch[0]) as {
      changed?: boolean;
      summary?: string;
      severity?: string;
      extracted_opportunities?: unknown[];
    };

    return {
      changed: parsed.changed ?? false,
      summary: parsed.summary ?? 'No summary provided',
      severity: parsed.severity ?? 'info',
      extractedOpportunities: parsed.extracted_opportunities ?? [],
      model: data.model ?? 'unknown',
      tokensUsed,
    };
  } catch (err) {
    // Claude analysis is best-effort — log and return null so the scout continues
    console.error('[source-scout] analyzeRegionWithClaude error:', err);
    return null;
  }
}

// ─── Tool definition ─────────────────────────────────────────────────

export const sourceScoutTool = defineTool<Input, Output>({
  name: 'finder.scout_source',
  namespace: 'finder',
  description:
    'Scout a source profile for changes. Fetches the page, compares against admin-annotated regions, and uses Claude to detect meaningful changes like new opportunities.',
  inputSchema: InputSchema,
  requiredRole: 'rfp_admin',
  tenantScoped: false,

  async handler(input, ctx) {
    const { sourceId } = input;
    const actorId = ctx.actor.id;
    const warnings: string[] = [];

    // ── 1. Load source profile ──────────────────────────────────────
    let profiles;
    try {
      profiles = await sql<
      {
        id: string;
        name: string;
        baseUrl: string;
        bookmarkUrl: string | null;
        isActive: boolean;
      }[]
    >`
      SELECT id, name, base_url, bookmark_url, is_active
      FROM source_profiles
      WHERE id = ${sourceId}::uuid
    `;
    } catch (err) {
      console.error('[finder.scout_source] profile query failed:', err);
      throw err;
    }

    if (profiles.length === 0) {
      throw new NotFoundError(`source profile not found: ${sourceId}`);
    }
    const profile = profiles[0];

    if (!profile.isActive) {
      warnings.push('Source profile is inactive');
    }

    // ── 2. Load active regions ──────────────────────────────────────
    let regions;
    try {
      regions = await sql<
      {
        id: string;
        name: string;
        selectorHint: string | null;
        contentContext: string | null;
        regionType: string;
        sampleText: string | null;
      }[]
    >`
      SELECT id, name, selector_hint, content_context, region_type, sample_text
      FROM source_regions
      WHERE profile_id = ${sourceId}::uuid
        AND is_active = true
      ORDER BY created_at
    `;
    } catch (err) {
      console.error('[finder.scout_source] regions query failed:', err);
      throw err;
    }

    if (regions.length === 0) {
      warnings.push(
        'No active regions defined for this source. Add regions with selector hints and content context to enable change detection.',
      );
      // Still fetch the page and create a whole-page snapshot
    }

    // ── 3. Fetch the page ───────────────────────────────────────────
    const fetchUrl = profile.bookmarkUrl ?? profile.baseUrl;
    let pageHtml: string;
    try {
      const result = await fetchPageHtml(fetchUrl);
      pageHtml = result.html;
      if (result.warning) {
        warnings.push(result.warning);
      }
    } catch (err) {
      if (err instanceof ToolExternalError) throw err;
      throw new ToolExternalError(
        `failed to fetch source page: ${err instanceof Error ? err.message : String(err)}`,
        { sourceId, url: fetchUrl },
      );
    }

    const fullPageText = htmlToText(pageHtml);

    // ── 4. Process each region ──────────────────────────────────────
    let snapshotsCreated = 0;
    let diffsFound = 0;
    let meaningfulChanges = 0;
    const regionResults: RegionDiff[] = [];
    const allExtractedOpps: unknown[] = []; // accumulated across regions → the candidate queue

    // If no regions, process the full page as a single implicit region
    const regionsToProcess =
      regions.length > 0
        ? regions
        : [
            {
              id: null as string | null,
              name: 'Full Page',
              selectorHint: null as string | null,
              contentContext: 'Full page content' as string | null,
              regionType: 'content',
              sampleText: null as string | null,
            },
          ];

    for (const region of regionsToProcess) {
      const regionText = fullPageText;
      const hash = contentHash(regionText);

      // Check last snapshot for this region
      let lastSnapshots;
      let newSnapshots;
      try {
        lastSnapshots = await sql<
          { id: string; contentHash: string; contentText: string | null }[]
        >`
          SELECT id, content_hash, content_text
          FROM source_snapshots
          WHERE profile_id = ${sourceId}::uuid
            AND (region_id = ${region.id}::uuid OR (region_id IS NULL AND ${region.id}::uuid IS NULL))
          ORDER BY captured_at DESC
          LIMIT 1
        `;
      } catch (err) {
        console.error('[finder.scout_source] snapshot lookup failed:', err);
        throw err;
      }

      const lastSnapshot = lastSnapshots[0] ?? null;
      const hashChanged = !lastSnapshot || lastSnapshot.contentHash !== hash;

      // Always create a new snapshot
      try {
        newSnapshots = await sql<{ id: string }[]>`
          INSERT INTO source_snapshots
            (profile_id, region_id, content_hash, content_text)
          VALUES
            (${sourceId}::uuid, ${region.id}::uuid, ${hash}, ${regionText.slice(0, 50000)})
          RETURNING id
        `;
      } catch (err) {
        console.error('[finder.scout_source] snapshot insert failed:', err);
        throw err;
      }
      snapshotsCreated++;
      const newSnapshotId = newSnapshots[0].id;

      // If hash unchanged, no diff needed
      if (!hashChanged) {
        regionResults.push({
          regionId: region.id ?? 'full-page',
          regionName: region.name,
          changed: false,
          summary: 'No changes detected (content hash match)',
          severity: 'info',
          extractedOpportunities: [],
        });
        continue;
      }

      // Hash changed — ask Claude to analyze
      const analysis = await analyzeRegionWithClaude(
        profile.name,
        region.name,
        region.contentContext ?? '',
        lastSnapshot?.contentText ?? region.sampleText,
        regionText,
        pageHtml,
      );

      const isMeaningful = analysis?.changed ?? hashChanged;
      const summary = analysis?.summary ?? 'Content hash changed but Claude analysis unavailable';
      const severity = analysis?.severity ?? 'low';
      const extractedOpps = analysis?.extractedOpportunities ?? [];

      // Create diff record
      try {
        await sql`
          INSERT INTO source_diffs
            (profile_id, region_id, prev_snapshot_id, next_snapshot_id,
             is_meaningful, summary, extracted_opportunities, severity,
             claude_model, claude_tokens_used)
          VALUES
            (${sourceId}::uuid, ${region.id}::uuid,
             ${lastSnapshot?.id ?? null}::uuid, ${newSnapshotId}::uuid,
             ${isMeaningful}, ${summary},
             ${sql.json((extractedOpps) as Parameters<typeof sql.json>[0])}, ${severity},
             ${analysis?.model ?? null}, ${analysis?.tokensUsed ?? null})
        `;
      } catch (err) {
        console.error('[finder.scout_source] diff insert failed:', err);
        throw err;
      }

      diffsFound++;
      if (isMeaningful) meaningfulChanges++;
      if (isMeaningful && extractedOpps.length > 0) allExtractedOpps.push(...extractedOpps);

      regionResults.push({
        regionId: region.id ?? 'full-page',
        regionName: region.name,
        changed: isMeaningful,
        summary,
        severity,
        extractedOpportunities: extractedOpps,
      });
    }

    // ── 5. Update source profile last_crawl_at ──────────────────────
    try {
      await sql`
        UPDATE source_profiles
        SET last_crawl_at = now(), updated_at = now()
        WHERE id = ${sourceId}::uuid
      `;
    } catch (err) {
      console.error('[finder.scout_source] profile update failed:', err);
      throw err;
    }

    // ── 6. Emit scout event ─────────────────────────────────────────
    await emitEventSingle({
      namespace: 'finder',
      type: 'source.scouted',
      actor: {
        type: ctx.actor.type,
        id: actorId,
        email: ctx.actor.email ?? undefined,
      },
      payload: {
        correlationId: randomUUID(),
        sourceId,
        sourceName: profile.name,
        snapshotsCreated,
        diffsFound,
        meaningfulChanges,
        warnings,
      },
    });

    // ── 7. If meaningful changes, emit change_detected ──────────────
    if (meaningfulChanges > 0) {
      await emitEventSingle({
        namespace: 'finder',
        type: 'source.change_detected',
        actor: {
          type: ctx.actor.type,
          id: actorId,
          email: ctx.actor.email ?? undefined,
        },
        payload: {
          correlationId: randomUUID(),
          sourceId,
          sourceName: profile.name,
          meaningfulChanges,
          regionResults: regionResults.filter((r) => r.changed),
        },
      });
    }

    // ── 8. Feed extracted opportunities into the candidate review queue ──
    // Materialize each Claude-extracted opportunity as a scout_findings row (classified NEW/UPDATE)
    // so the HITL scout and the crawler feed ONE review→release queue. Best-effort — a failure here
    // must never fail the scout run.
    let candidatesQueued = 0;
    if (allExtractedOpps.length > 0) {
      try {
        const { materializeExtractedOpportunities } = await import('@/lib/scout/candidates');
        candidatesQueued = await materializeExtractedOpportunities(
          sourceId,
          profile.name,
          allExtractedOpps,
          { actorId, actorEmail: ctx.actor.email ?? null },
        );
      } catch (err) {
        console.error('[finder.scout_source] materialize candidates failed:', err);
      }
    }

    ctx.log?.info?.({
      msg: 'finder.scout_source completed',
      sourceId,
      snapshotsCreated,
      diffsFound,
      meaningfulChanges,
      candidatesQueued,
    });

    return {
      sourceId,
      sourceName: profile.name,
      snapshotsCreated,
      diffsFound,
      meaningfulChanges,
      regionResults,
      warnings,
    };
  },
});
